ledger-service/http-json/src/test/scala/com/digitalasset/http/Consume.scala

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.http
+
+import akka.stream.scaladsl.Sink
+import org.scalactic.source
+import scalaz.{~>, Functor}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/** This example script reads two elements unconditionally, checking whether
+  * the first element is "good" and that the second element has `true` as its
+  * first member, then performs a future, then drains any remaining elements
+  * from the stream, yielding a result that uses elements of all three reads.
+  *
+  * {{{
+  *   val mySyntax = Consume.syntax[Foo] // Foo = stream element type
+  *   import mySyntax._
+  *
+  *   val sink: Sink[Foo, Future[Bar]] = Consume.interpret(for {
+  *     a <- readOne
+  *     if aGoodFoo(a) // abort if false
+  *
+  *     Foo(true, b) <- readOne // abort if pattern doesn't match
+  *     _ <- liftF(Future { q(a, b) /* complete this future before continuing */ })
+  *
+  *     rest <- drain // Seq[Foo] of remainder; use only if you expect more elements
+  *   } yield Bar(a, b, rest)
+  * }}}
+  */
+sealed abstract class Consume[-T, +V]
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+object Consume {
+  import scalaz.Free
+  import scalaz.std.scalaFuture._
+
+  type FCC[T, V] = Free[Consume[T, ?], V]
+  type Description = source.Position
+  final case class Listen[-T, +V](f: T => V, desc: Description) extends Consume[T, V]
+  final case class Drain[S, -T, +V](init: S, next: (S, T) => S, out: S => V) extends Consume[T, V]
+  final case class Emit[+V](run: Future[V]) extends Consume[Any, V]
+
+  /** Strictly speaking, this function really returns the following, these
+    * are just passed to `Sink.foldAsync` for convenience:
+    *
+    * {{{
+    *   {
+    *     type O
+    *     val init: O
+    *     val step: (O, T) => Future[O]
+    *     val out: O => Future[V]
+    *   }
+    * }}}
+    */
+  def interpret[T, V](steps: FCC[T, V])(implicit ec: ExecutionContext): Sink[T, Future[V]] =
+    Sink
+      .foldAsync(steps) { (steps, t: T) =>
+        // step through steps until performing exactly one listen,
+        // then step through any further steps until encountering
+        // either the end or the next listen
+        def go(steps: FCC[T, V], listened: Boolean): Future[FCC[T, V]] =
+          steps.resume fold ({
+            case listen @ Listen(f, _) =>
+              if (listened) Future successful (Free roll listen) else go(f(t), true)
```

Comment:
I considered whether `Free` was really appropriate here, and it ultimately came down to things like this. Interpreting these scripts must pause when reaching a `readOne` (`Listen` here). However, a naive interpretation will interpret a `flatMap`

```scala
  [E] (first: Consume[T, E], next: E => Consume[T, V]) => ...
```

by recursive descent into `first`. However, should it encounter `Listen` on this recursive descent, what then? Assuming it's the second listen encountered, it must immediately stop, returning it _and all the remaining actions_, including the ones unpacked as we recur.

By contrast, in this code, `f: T => Free[Consume[T, ?], V]`, where the function result represents the complete remainder of the sequence, _regardless of how arbitrarily nested within `first`s this `Listen` was_. That is the result of `resume`; in other words, the key feature we needed was turning every left-associated bind `fx >== k >== t` into a right-associated bind `fx >== (\x -> k x >== t)`, exactly what `Free` and `resume` are designed for, thus permitting scripts to be factored in a natural way.